### PR TITLE
Adding info on migration from shinytest to shinytest2

### PR DESCRIPTION
**Pull request by jen-machin**
Now that Rich has updated the dashboard page, I added in the info on migrating from shinytest to shinytest2 😄 I had previously done this on another branch but the structure of the pages changed so much that it was easier to just create a new branch and start again to avoid chaos!!

### CODE REVIEW

File: writing-visualising/dashboards_rshiny.qmd [hunk 1]
```diff
@@ -490,6 +496,43 @@ The expected outputs are saved in a [_snaps/shinytest2](https://github.com/dfe-a
...
+
+There is [guidance on migrating to shinytest2](https://rstudio.github.io/shinytest2/articles/z-migration.html) available. To complete the migration, run `shinytest2::migrate_from_shinytest(PATH_TO_APP)`, making sure to include the full file path to where your project is saved, e.g. `shinytest2::migrate_from_shinytest('C:/R projects/analysts-guide')`. If you copy your file path from Windows Explorer, remember to change the `\` to `/` otherwise you'll encounter an error.
+
+The migrate_from_shinytest function will search for your "tests" folder, and then look inside the folder for any scripts containing UI tests. A new `testthat` folder will be automatically created, containing the code that runs your tests, as well as the snapshots generated when your tests are run. Don't panic if your old test script is deleted as part of the migration, this is supposed to happen! It is recommended that someone who was familiar with the original tests completes QA of the new tests to make sure that everything looks as expected. 
+
+There are a couple of potential issues that you may run into when completing migration or running your tests:
+
+**Chromote**
```
Comment (cjrace) [inline writing-visualising/dashboards_rshiny.qmd:514]: Should this be a heading? (I'd assume h4?)

File: writing-visualising/dashboards_rshiny.qmd [hunk 2]
```diff
@@ -490,6 +496,43 @@ The expected outputs are saved in a [_snaps/shinytest2](https://github.com/dfe-a
...
+
+Shinytest2 makes use of the Chromote package, which needs to connect to your local instance of Google Chrome. Please make sure you have Google Chrome installed before you attempt to use shinytest2, otherwise it will not work. Sometimes your copy of Chrome might be installed in a different place to where Chromote expects to find it, which means that your tests will not be able to run. There is information on how to resolve this issue [in the Tips for Using R section](../learning-development/r.html#shinytest2---invalid-path-to-chrome-error-message)
+
+**Formatting of line endings**
```
Comment (cjrace) [inline writing-visualising/dashboards_rshiny.qmd:518]: Should this be a heading? (I'd assume h4?)

File: writing-visualising/dashboards_rshiny.qmd [hunk 3]
```diff
@@ -490,6 +496,43 @@ The expected outputs are saved in a [_snaps/shinytest2](https://github.com/dfe-a
...
+
+The migration function will only work if specific line endings (also known as carriage returns) are used in your code. It will not be immediately obvious to you what kind of line endings your code uses as default. If you receive an error that looks like this, or any error that seems to be pointing at the blank space at the end of a line of code, then you may be having this issue:
+
+```
```
Comment (cjrace) [inline writing-visualising/dashboards_rshiny.qmd:522]: I don't think this code block is formatting right, should it have highlighting and a little copy symbol in the top right corner? [image from local build on this branch]
image

File: writing-visualising/dashboards_rshiny.qmd [hunk 4]
```diff
@@ -490,6 +496,43 @@ The expected outputs are saved in a [_snaps/shinytest2](https://github.com/dfe-a
...
+
+There are a couple of potential issues that you may run into when completing migration or running your tests:
+
+**Chromote**
+
+Shinytest2 makes use of the Chromote package, which needs to connect to your local instance of Google Chrome. Please make sure you have Google Chrome installed before you attempt to use shinytest2, otherwise it will not work. Sometimes your copy of Chrome might be installed in a different place to where Chromote expects to find it, which means that your tests will not be able to run. There is information on how to resolve this issue [in the Tips for Using R section](../learning-development/r.html#shinytest2---invalid-path-to-chrome-error-message)
```
Comment (cjrace) [inline writing-visualising/dashboards_rshiny.qmd:516]: Scope creep a little - but in that section you refer to here we don't put anything on how to find where your chrome is installed. I wonder if someone has this issue, then it might be worth adding a quick paragraph on how to find that out? (I'm assuming there's a quick way like hit windows and right click chrome to find it or a command you could run or something to show you what path it should be?)

**Review by cjrace — CHANGES_REQUESTED** [review]

File: writing-visualising/dashboards_rshiny.qmd (hunk 4 above)
Comment (jen-machin) [reply writing-visualising/dashboards_rshiny.qmd:516]: Forgot the command doesn't always work properly so I added a different method of finding out!

**Comment by jen-machin** [comment]
Also went totally off tangent and fixed a load of problems with formatting in code blocks and updated the rounding section to include more info (I should have done this in a separate branch, oopsy)

File: learning-development/r.qmd [hunk 5]
```diff
@@ -92,15 +92,15 @@ Sometimes renv::restore() can fail, and when in specific renv-controlled project
...
+``` {r proxy_settings, eval=FALSE}
 Sys.setenv(no_proxy="*") 
 
 ```
 
 
 2. Specifying the renv library as the install location. It's a bit of a fudge, though these lines are helpful to get the packages from the renv lockfile installed and you running the project when needed:
-
-```
+
```
Comment (cjrace) [inline learning-development/r.qmd:102]: Looks like stray whitespace?

File: renv.lock
Comment (cjrace) [inline renv.lock:1]: Looks like the versions of some packages have gone backwards - was this intentional?

File: learning-development/r.qmd (hunk 5 above)
Comment (jen-machin) [reply learning-development/r.qmd:102]: I don't even remember editing this line 🤣

File: renv.lock
Comment (jen-machin) [reply renv.lock:1]: It was not, renv hates me

File: learning-development/r.qmd [hunk 6]
```diff
@@ -430,9 +430,20 @@ This can be done most easily by navigating in RStudio through `Tools > Global op
...
+However, when investigating numbers in R, you can see that they are not always stored in the way you might think. This is due to floating point numbers. Whilst humans use base 10 (numbers 0-9), computers store numbers in a binary (base 2) system as a combinations of 1s and 0s. When translating between base 2 and base 10, sometimes numbers can be not quite what we expect. For example, asking R to print the number 1.4 to 18 decimal places shows that it is actually stored as:
 
+``` {r round_floating_point, eval=FALSE}
+sprintf("%.18f", 1.4)
+[1] "1.399999999999999911"
 ```
+
+These slight differences in the floating point numbers behind the scenes can result in unusual rounding errors using round(). If you're interested, [there's more information available in these forum posts](https://stat.ethz.ch/pipermail/r-help/2007-August/138014.html). 
+ 
+To round numbers upwards in a way that is consistent with SQL and other programming languages, you can create a custom function like the one below:
```
Comment (rmbielby) [inline learning-development/r.qmd:444]: I think we should switch this up and recommend using the already created function from dfeR, rather than getting everyone to reproduce the custom function.

So switch this line to "To round numbers upwards..., we recommend using the roundFiveUp() function provided by the dfeR package." Could also potentially put roundFiveUp() as a link to the code in dfeR instead of recreating it here.

**Review by rmbielby — CHANGES_REQUESTED** [review]

**Review by rmbielby — APPROVED** [review]

**Review by cjrace — APPROVED** [review]
Have made one tiny tiny change to fix the whitespace while I was running it locally already, happy to approve .

Double check that last commit of mine and then I think it's good to merge 😄

**Review by rmbielby — APPROVED** [review]
